src/NUnitEngine/nunit.engine/Services/TestSelectionParser.cs

```diff
@@ -0,0 +1,275 @@
+            }
+
+            if (terms.Count == 1)
+                return terms[0];
+
+            var sb = new StringBuilder("<or>");
+
+            foreach (string term in terms)
+                sb.Append(term);
+
+            sb.Append("</or>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single element or an and expression and return the xml
+        /// </summary>
+        public string ParseFilterTerm()
+        {
+            var elements = new List<string>();
+            elements.Add(ParseFilterElement());
+
+            while (true)
+            {
+                Token op = Peek();
+                if (op != AND_OP1 && op != AND_OP2)
+                    break;
+
+                NextToken();
+                elements.Add(ParseFilterElement());
+            }
+
+            if (elements.Count == 1)
+                return elements[0];
+
+            var sb = new StringBuilder("<and>");
+
+            foreach (string element in elements)
+                sb.Append(element);
+
+            sb.Append("</and>");
+
+            return sb.ToString();
+        }
+
+        /// <summary>
+        /// Parse a single filter element such as a category expression
+        /// and return the xml representation of the filter.
+        /// </summary>
+        public string ParseFilterElement()
+        {
+            Token nextOp = Peek();
+            if (nextOp == LPAREN || nextOp == NOT_OP)
+                return ParseExpressionInParentheses();
+
+            string lhs = GetWord();
+
+            switch (lhs)
+            {
+                case "id":
+                    string op = GetOperator("=", "==");
+                    string values = GetIdList();
+                    return string.Format("<id>{0}</id>", values);
+
+                case "cat":
+                case "category":
+                case "method":
+                case "class":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = GetWordList();
+                    string tag = lhs == "category" ? "cat" : lhs;
+                    switch (op)
+                    {
+                        case "=":
+                        case "==":
+                            return string.Format("<{0}>{1}</{0}>", tag, values);
+                        case "!=":
+                            return string.Format("<not><{0}>{1}</{0}></not>", tag, values);
+                        default:
+                            return string.Format("<{0} op='{1}'>{2}</{0}>", tag, op, values);
+                    }
+
+                case "name":
+                case "test":
+                case "fullname":
+                    op = GetOperator("=", "==", "!=", "=~", "!~");
+                    values = Peek().Kind == TokenKind.String
+                        ? GetString()
+                        : GetWord();
+                    tag = lhs == "fullname" ? "test" : lhs;
```

Comment:
So, name is the name we generate and is mainly useful for things like TestCases?

I am fine with method/class. test and fullname don't really work for me though. Test sort of makes sense because it mimics the old command line, but fullname isn't really correct either. It isn't actually the fullname, it is all tests contained within the given name be it a namespace, class or right down to the method. I am drawing a blank on what to call that though :frowning: 

Do we also want a namespace keyword to run all tests within a namespace?

I also think class should work on the fully qualified classname or just the simple classname. I always hate fully qualifying names with the --test command. I should only have to do so if there are multiple classes with the same name in different namespaces.